provider/azure/internal/imageutils/images.go

```diff
@@ -159,22 +178,38 @@ func legacyUbuntuSKU(ctx context.ProviderCallContext, base jujubase.Base, stream
 	if stream == dailyStream {
 		offer = fmt.Sprintf("%s-daily", offer)
 	}
-	desiredSKUPrefix := strings.ReplaceAll(base.Channel.Track, ".", "_")
 
-	logger.Debugf("listing SKUs: Location=%s, Publisher=%s, Offer=%s", location, ubuntuPublisher, offer)
+	logger.Debugf(
+		"listing SKUs: Base=%s, Series=%s, Location=%s, Stream=%s, Publisher=%s, Offer=%s",
+		base.Channel.Track, series, location, stream, ubuntuPublisher, offer,
+	)
 	result, err := client.ListSKUs(ctx, location, ubuntuPublisher, offer, nil)
 	if err != nil {
 		return "", "", errorutils.HandleCredentialError(errors.Annotate(err, "listing Ubuntu SKUs"), ctx)
 	}
-	for _, img := range result.VirtualMachineImageResourceArray {
-		skuName := *img.Name
-		logger.Debugf("found Azure SKU Name: %v", skuName)
-		if !strings.HasPrefix(skuName, desiredSKUPrefix) {
-			logger.Debugf("ignoring SKU %q (does not match series %q)", skuName, series)
-			continue
+
+	skuName, err := selectUbuntuSKULegacy(base, series, stream, result.VirtualMachineImageResourceArray, preferGen1Image)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	return skuName, offer, nil
+}
+
+func selectUbuntuSKULegacy(
+	base jujubase.Base, series, stream string,
+	images []*armcompute.VirtualMachineImageResource, preferGen1Image bool,
+) (string, error) {
+	// We prefer to use v2 SKU if available. But if not, we fall back to v1 SKU.
```

Comment:
The above comment is not strictly true - whether we want gen 1 or gen2 is determined by the function arg

Instead of passing that arg and having to make 2 api calls to query the image catalog, wouldn't it be better to return a struct with a list of gen 1 images and a list of gen 2 images. Then the caller can look for gen 2 or gen 1 without needing multiple api calls.

Reply:
This PR focuses on fixing the image lookup bug. 
To achieve the single API call, we need to change the StartInstance in the env which has retry logic for different retriable errors.
It could be a refactor for the future but not in this PR.